[runtime asserts] deduplicate runtime asserts & CSE

pianpwk opened the pull request:
This PR adds deduplication and CSE for runtime asserts. Existing size computation in the graph is CSE'd along with added runtime asserts, and redundant asserts are removed. Shape calls on intermediate tensors are also turned into compute on input sizes if possible, allowing intermediate tensors to be freed earlier. For example:
```
z = torch.cat([x, x], dim=0)  # 2*s0
w = z.repeat(y.shape[0])  # 2*s0*s1
_w = w.shape[0]
# something with _w ...

# turns into ->
s0 = x.shape[0]
s1 = y.shape[0]
_w0 = 2 * s0
_w = _w0 * s1
```

Additionally, constrain_range calls are deduplicated. Single-symbol bound checks for unbacked symbols (e.g. u0 >= 0, u0 <= 5) and sym_constrain_range.default calls are also removed, since they accumulate range info in the ShapeEnv, and are replaced with two _assert_scalar.default calls that check the min/max bounds. For example:
```
torch.sym_constrain_range_for_size(n, min=2, max=16)
torch.sym_constrain_range(n, min=4, max=20)
torch._check(n >= 0)
torch._check(n >= 3)
torch._check(n <= 14)

# turns into
torch.sym_constrain_range_for_size(n)
torch._check(n >= 4)
torch._check(n <= 14)
```



cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

pianpwk committed:
lifted constants

pianpwk committed:
Merge branch 'main' of https://github.com/pytorch/pytorch

pianpwk committed:
Merge branch 'main' of https://github.com/pytorch/pytorch

pianpwk committed:
Merge branch 'main' of https://github.com/pytorch/pytorch

pianpwk committed:
test

pianpwk committed:
test

pianpwk committed:
typo

pianpwk committed:
stash

pianpwk committed:
changes

pianpwk committed:
merge

pianpwk committed:
add

pianpwk committed:
remove comments

pianpwk committed:
comment

pianpwk committed:
fix canonicalize

pianpwk committed:
tests

pianpwk committed:
fix explicit guard test

pianpwk committed:
Update runtime_assert.py

pianpwk commented:
one exported program error: if the output node is CSE'd the graph signature changes -> will fix this

pianpwk committed:
fix broken added test

pianpwk committed:
dce

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -226,9 +227,11 @@ def forward(self, x):
-        with self.assertRaisesRegex(
-            RuntimeError, r"Invalid value range for 3 between \[4, 5\]"
-        ):
+        if is_non_strict_test(self._testMethodName):
```
Hmm why does it change?

pianpwk replied:
Original (main):
```
ExportedProgram:
    class GraphModule(torch.nn.Module):
        def forward(self, x: "i64[1]"):
             # File: /data/users/pianpwk/copy/pytorch/test/export/test_export.py:220 in forward, code: b = x.item()
            item: "Sym(u0)" = torch.ops.aten.item.default(x);  x = None
            
             # 
            sym_constrain_range_for_size_default = torch.ops.aten.sym_constrain_range_for_size.default(item)
            
            # No stacktrace found for following nodes
            sym_constrain_range = torch.ops.aten.sym_constrain_range.default(item, min = 4, max = 5)
            ge: "Sym(u0 >= 0)" = item >= 0
            _assert_scalar = torch.ops.aten._assert_scalar.default(ge, "Runtime assertion failed for expression 0 <= u0 on node 'ge_1'");  ge = None
            ge_1: "Sym(u0 >= 4)" = item >= 4
            _assert_scalar_1 = torch.ops.aten._assert_scalar.default(ge_1, "Runtime assertion failed for expression 4 <= u0 on node 'ge_2'");  ge_1 = None
            le: "Sym(u0 <= 5)" = item <= 5
            _assert_scalar_2 = torch.ops.aten._assert_scalar.default(le, "Runtime assertion failed for expression u0 <= 5 on node 'le_1'");  le = None
            
             # File: /data/users/pianpwk/copy/pytorch/test/export/test_export.py:224 in forward, code: return torch.full((b, 1), 1)
            full: "i64[u0, 1]" = torch.ops.aten.full.default([item, 1], 1, device = device(type='cpu'), pin_memory = False);  item = None
            return (full,)
```

Current (strict-mode): the assert_async calls are added separately, and should be removed and the pass deprecated soon: https://github.com/pytorch/pytorch/pull/125948. For non-strict we don't run this pass.

```
ExportedProgram:
    class GraphModule(torch.nn.Module):
        def forward(self, x: "i64[1]"):
             # File: /data/users/pianpwk/pytorch/test/export/test_export.py:221 in forward, code: b = x.item()
            item: "Sym(u0)" = torch.ops.aten.item.default(x);  x = None
            
             # 
            ge: "Sym(u0 >= 4)" = item >= 4
            scalar_tensor_default: "f32[]" = torch.ops.aten.scalar_tensor.default(ge);  ge = None
            _assert_async_msg = torch.ops.aten._assert_async.msg(scalar_tensor_default, 'item is outside of inline constraint [4, 5].');  scalar_tensor_default = None
            le: "Sym(u0 <= 5)" = item <= 5
            scalar_tensor_default_1: "f32[]" = torch.ops.aten.scalar_tensor.default(le);  le = None
            _assert_async_msg_1 = torch.ops.aten._assert_async.msg(scalar_tensor_default_1, 'item is outside of inline constraint [4, 5].');  scalar_tensor_default_1 = None
            
             # File: /data/users/pianpwk/pytorch/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:158 in call, code: node.meta["stack_trace"] = "".join(traceback.format_stack(limit=1))
            sym_constrain_range_for_size = torch.ops.aten.sym_constrain_range_for_size.default(item, min = 4, max = 5)
            
             # File: /data/users/pianpwk/pytorch/test/export/test_export.py:225 in forward, code: return torch.full((b, 1), 1)
            full: "i64[u0, 1]" = torch.ops.aten.full.default([item, 1], 1, device = device(type='cpu'), pin_memory = False);  item = None
            return (full,)
```

Current (non-strict), ideal:
```
ExportedProgram:
    class GraphModule(torch.nn.Module):
        def forward(self, x: "i64[1]"):
             # File: /data/users/pianpwk/pytorch/test/export/test_export.py:221 in forward, code: b = x.item()
            item: "Sym(u1)" = torch.ops.aten.item.default(x);  x = None
            
             # 
            sym_constrain_range_for_size_default = torch.ops.aten.sym_constrain_range_for_size.default(item, min = 4, max = 5)
            
             # File: /data/users/pianpwk/pytorch/test/export/test_export.py:225 in forward, code: return torch.full((b, 1), 1)
            full: "i64[u1, 1]" = torch.ops.aten.full.default([item, 1], 1, device = device(type='cpu'), pin_memory = False);  item = None
            return (full,)
```

pianpwk replied:
Others are a similar story, should be unified b/w strict & non-strict once the above PR is merged

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -2812,13 +2822,14 @@ def forward(self, x):
-            RuntimeError,
-            r"Invalid value range for 30 between \[4, 7\]",
-        ) as cm:
+        if is_non_strict_test(self._testMethodName):
```
This feels like we are dropping asserts somewhere?

tugsbayasgalan commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -27,11 +37,167 @@ def _get_example_value(node: fx.Node) -> Optional[str]:
+        val = node.meta["example_value"]
     elif "val" in node.meta:
-        return node.meta["val"]
+        val = node.meta["val"]
```
seems like not needed?

tugsbayasgalan commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -347,51 +373,573 @@ def go(node, keypath):
+            hash_node = hash_cons[sym_shape].node
+            node.replace_all_uses_with(hash_node)
+            log.debug("CSEd intermediate size node for expr %s", sym_shape)
+            _dce_from_node(node)  # here we should DCE
```
Why can't we just call DCE all the way in the end?

pianpwk replied:
We can, this just doesn't requiring iterating over all the nodes to do so, since we know what to DCE on. But maybe that is cleaner..

pianpwk committed:
Merge branch 'main' of https://github.com/pytorch/pytorch into pianpwk/dedup_runtime_asserts

pianpwk committed:
fix test failures

pianpwk committed:
add missing files

pianpwk committed:
unbacked_interp()

pianpwk commented:
The export test `test_le_ge_ne_eq` I added seems flaky even locally, the assert_scalar count differs from run to run (will debug this), and some failures on op counts in `test/dynamo/test_export.py`

ezyang commented on torch/utils/_sympy/expr_ranges.py:
```diff
@@ -0,0 +1,178 @@
+    Refinement calls return None if the call had no effect, and a tuple of (SymRel, val)
+    if the range is modified. The returned SymRel may differ from the called method,
+    for example lt(3, strict=False), gt(3, strict=False) returns (SymRel.EQ, 3)
+    since the range is made static.
```
I'm a bit unsure why this concept is needed at the moment.

lezcano replied:
Isn't this `ValueRanges.__and__`?

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -847,10 +847,11 @@ def _constrain_range_for_size(a, min: Optional[int] = None, max: Optional[int] =
+        isinstance(a, SymInt)
+        and isinstance(a.node.expr, sympy.Symbol)
+    ):  # follow _advise_is_size(), no-op if a is concrete value
+        a.node.shape_env._constrain_range_for_size(a.node.expr, min, max)
```
While this probably is OK, it is a bit suspicious that you needed to no-op the asserts, because it suggests that maybe you might be using the wrong API downstream.

ezyang commented on torch/utils/_sympy/interp.py:
```diff
@@ -110,9 +113,15 @@ def sympy_interp(
     expr: Union[sympy.Expr, SympyBoolean],
+    hash_cons: Optional[Dict[sympy.Expr, fx.Node]] = None,
     *,
+    insert_after_args=False,
```
I'm not sure why this is needed. In particular, we visit each node before all of its uses, so on the outer scope you simply need an "insert before" context manager, where before is the node immediately after the insertions we want to do, and the insertion point will appropriately move forward as you insert more nodes.

ezyang commented on torch/utils/_sympy/interp.py:
```diff
@@ -110,9 +113,15 @@ def sympy_interp(
     analysis,
     env: Dict[sympy.Symbol, Any],
     expr: Union[sympy.Expr, SympyBoolean],
+    hash_cons: Optional[Dict[sympy.Expr, fx.Node]] = None,
```
It's probably possible to do the hash cons directly in the analysis itself, rather than backing it into the interp, but it's probably not a huge deal to have it here either. However, the typing here is overly restrictive: the return type of the hash cons is just whatever the analysis returns.

lezcano replied:
I'd vote against modifying this file. This implements a SymPy interpreter. The new changes should be done on top of it, by inheritance or any other composition mechanism. There is no reason for the SymPy interpreter to depend on `fx.Node`...

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -1,7 +1,11 @@
 # mypy: allow-untyped-defs
+import contextlib
```
Is attempting to look at the diff here at all useful, or should I just read it as an entirely new file?

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -34,6 +44,161 @@ def _get_example_value(node: fx.Node) -> Optional[str]:
+            return val.node.expr
+        elif isinstance(val, sympy.Expr):
+            return val
+    return None
```
It's not obvious to me why this is needed, we shouldn't ever have sympy.Expr stored in example_value

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -34,6 +44,161 @@ def _get_example_value(node: fx.Node) -> Optional[str]:
+
+def _canonicalize_expr(expr, mapping):
+    """
+    https://github.com/pytorch/pytorch/pull/128411 means we won't get expressions like -u0 <= 0
```
cc @lezcano

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -34,6 +44,161 @@ def _get_example_value(node: fx.Node) -> Optional[str]:
+    """
+    https://github.com/pytorch/pytorch/pull/128411 means we won't get expressions like -u0 <= 0
+    for calls like torch._check(-u0 <= 0) anymore, but it does seem like we can get 4 <= u0, or u0 < s0,
+    which are slightly difficult to handle when we want to track ranges.
```
Yeah, I still don't understand why you need expr ranges......

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -34,6 +44,161 @@ def _get_example_value(node: fx.Node) -> Optional[str]:
+    return (
+        node.op == "call_function"
+        and node.target in [torch.ops.aten.sym_size.int, "size"]
+    )
```
This is kinda sus, why is there a helper specifically for size but not, e.g., for stride/storage offset

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -34,6 +44,161 @@ def _get_example_value(node: fx.Node) -> Optional[str]:
+                and not n.users
+                and n.op != "placeholder"
+            )
+        )
```
This looks unsafe. Because runtime assertion insertion can happen in graphs that haven't been functionalized, you cannot in general DCE arbitrary Tensor compute, because that computation may have side effects. You can only DCE "size-like" compute on SymInt/int, which is guaranteed to be purely functional as int is immutable.

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -48,35 +213,56 @@ def insert_deferred_runtime_asserts(
     tensor propagation, so we cannot conveniently insert them into the FX graph
     when they occur.  Instead, we accumulate them in the ShapeEnv, and in this
     pass insert them into the graph as proper tests.
```
This doc is almost definitely wrong now

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -85,16 +271,41 @@ def insert_deferred_runtime_asserts(
 
-    # TODO: Request simplification on runtime asserts before emitting them
+    RUNTIME_ASSERT_OPS = [
+        torch.ops.aten._assert_async.msg,
```
This is a kind of strange thing to handle here, since assert async operates on Tensor and we really shouldn't be touching Tensor compute in this pass

pianpwk replied:
I see, we'd been using it for the previous runtime asserts pass which we're deprecating, so I thought it was basically assert_scalar.

ezyang replied:
Ah yes, if you wanted to peephole "scalar compute -> make a tensor -> assert async" that's... reasonable, I guess. Better not to generate code like that in the first place though.

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -103,211 +314,41 @@ def insert_deferred_runtime_asserts(
+            if node.target in (
+                torch.ops.aten.sym_constrain_range_for_size.default,
+                torch._check_is_size,
+                torch.ops.aten.sym_constrain_range.default,  # technically not an is_size call, but doesn't take non-symbols
```
I don't understand. Then what the heck is an `is_size_call`?

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -347,51 +388,547 @@ def go(node, keypath):
+            sym_expr = expr
+
+        if is_size:  # make size-like
+            sym_expr_to_ranges[expr].set_is_size()
```
All of this seems unnecessarily complicated. I think it would be better to do a version of this pass that is not trying to do nontrivial semantic deduplication. So if I had `u0 > 0` vs `u0 >= 1` as runtime asserts waiting in the ShapeEnv, I wouldn't try to deduplicate these. It's the responsibility for the ShapeEnv to remove duplicates.

So in this regime, there are only two things we need to track: is the symbol size-like (because that's special), and then otherwise a set of expressions which evaluate to True in `_assert_scalar`. We do not even need `sym_constrain_range`, because we can simply desugar this to `low <= u0` and `u0 <= upper`. And because we are CSE'ing, we get some mild semantic deduplication, which is that if I have an RA saying `low <= u0` and I also have a VR saying `low <= u0`, I will end up only generating the `_assert_scalar` once.

lezcano commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -34,6 +44,161 @@ def _get_example_value(node: fx.Node) -> Optional[str]:
+    # try flipping, won't hurt
+    # can return if original expr is something like 4 <= s0
+    lhs, rhs = expr.args
+    if isinstance(expr, sympy.LessThan):
+        expr = sympy.GreaterThan(rhs, lhs)
+    elif isinstance(expr, sympy.StrictLessThan):
+        expr = sympy.StrictGreaterThan(rhs, lhs)
+    elif isinstance(expr, sympy.GreaterThan):
+        expr = sympy.LessThan(rhs, lhs)
```
This operation is not idempotent?? Why do you always flip the expresion from `a > b` into `b < a` AND from `b < a` into `a > b`??

lezcano replied:
This is not very canonical...
You don't need to reinvent the wheel here. If you want to find the additive constants in an expression, you should call canonicalize and find it filtering the args on the left / right of the expression.
Canonicalisation assures that the expression is just of the form `<` or `<=`, and that the additive constant is positive. You can then use this however you want in your bounds.

pianpwk committed:
merge

pianpwk committed:
revert changes

pianpwk committed:
reduce scope

pianpwk committed:
reduce scope

pianpwk commented on test/dynamo/test_logging.py:
```diff
@@ -687,6 +687,7 @@ def fn(a):
     "recompiles_verbose",
     "graph_breaks",
     "graph",
+    "graph_code",
```
additional logging count because runtime_asserts pass now runs (for CSE) even if runtime asserts are empty

pianpwk commented on test/dynamo/test_misc.py:
```diff
@@ -371,7 +370,6 @@ def unpack4(a, b):
             unpack4,
             2,
             expected_ops=5,
-            expected_ops_dynamic=ifdynstaticdefault(5, 7),
```
Previous graph had
```
size = a.size()
getitem_2 = size[0]
getitem_3 = size[1];  size = None
```

size[0] is 5, and size[1] is s0, which is a graph placeholder, so both calls get removed (later in dynamo?)

pianpwk commented on test/dynamo/test_misc.py:
```diff
@@ -926,7 +923,7 @@ def fn(x):
 
         torch._dynamo.testing.standard_test(
-            self, fn, 1, expected_ops=1, expected_ops_dynamic=ifdynstaticdefault(1, 10)
+            self, fn, 1, expected_ops=1, expected_ops_dynamic=ifdynstaticdefault(1, 8)
```
p += 2, p -= 2 result gets hash consed, and add node probably gets DCE'd?

pianpwk commented on test/dynamo/test_misc.py:
```diff
@@ -1382,14 +1379,14 @@ def _fn(a, b, func=func):
             get_test_fn(func=max),
             2,
             expected_ops=1,
-            expected_ops_dynamic=ifdynstaticdefault(1, 17),
```
Test case contains:
```
lst = [a.shape[0] + 1, 8, a.shape[0]]
x = func(lst)
y = func(iter(lst))
z = func(tuple(lst))
w = func(*lst)
return a + (x + y + z + w)
```
where func is max.

here the ShapeEnv seems to say a.shape[0] has range [7, inf], so all size calls resolve to s0 + 1, and get hash consed. 8 sym_max calls and 2 (size() + getitem) calls are removed.

pianpwk commented on test/dynamo/test_misc.py:
```diff
@@ -1928,7 +1925,7 @@ def fn(a):
         # expect 4 add / subs for static, 4 * 3 (size, index, math op) for dynamic
         torch._dynamo.testing.standard_test(
-            self, fn, 1, expected_ops=4, expected_ops_dynamic=ifdynstaticdefault(4, 12)
+            self, fn, 1, expected_ops=4, expected_ops_dynamic=ifdynstaticdefault(4, 6)
```
3 size() + 3 getitem(size, 0) calls get deduped

pianpwk commented on test/dynamo/test_repros.py:
```diff
@@ -1238,7 +1238,7 @@ def test_longformer_chunk(self):
             if torch._dynamo.config.automatic_dynamic_shapes:
                 self.assertExpectedInline(cnt.frame_count, """2""")
-                self.assertExpectedInline(cnt.op_count, """14""")
+                self.assertExpectedInline(cnt.op_count, """9""")
```
size calls get deduped, stride calls become compute on input symints

before:
```
def forward(self, s0 : torch.SymInt, s1 : torch.SymInt, L_hidden_states_ : torch.Tensor):
    l_hidden_states_ = L_hidden_states_
    size = l_hidden_states_.size(0)
    size_1 = l_hidden_states_.size(2)
    hidden_states = l_hidden_states_.view(size, 8, 512, size_1);  l_hidden_states_ = size = size_1 = None
    size_2 = hidden_states.size()
    getitem = size_2[0]
    getitem_3 = size_2[3];  size_2 = None
    stride = hidden_states.stride()
    getitem_4 = stride[0]
    getitem_5 = stride[1]
    getitem_6 = stride[2];  stride = None
    floordiv = getitem_5 // 2;  getitem_5 = None
    as_strided = hidden_states.as_strided(size = [getitem, 15, 512, getitem_3], stride = [getitem_4, floordiv, getitem_6, 1]);  hidden_states = getitem = getitem_3 = getitem_4 = floordiv = getitem_6 = None
    return (as_strided,)
```

after:
```
def forward(self, s0 : torch.SymInt, s1 : torch.SymInt, L_hidden_states_ : torch.Tensor):
    l_hidden_states_ = L_hidden_states_
    size = l_hidden_states_.size(0)
    size_1 = l_hidden_states_.size(2)
    hidden_states = l_hidden_states_.view(size, 8, 512, size_1);  l_hidden_states_ = size = size_1 = None
    mul = 4096 * s1
    mul_1 = 512 * s1
    floordiv = mul_1 // 2;  mul_1 = None
    as_strided = hidden_states.as_strided(size = [s0, 15, 512, s1], stride = [mul, floordiv, s1, 1]);  hidden_states = s0 = s1 = mul = floordiv = None
    return (as_strided,)
```

pianpwk commented on torch/utils/_sympy/interp.py:
```diff
@@ -178,3 +153,33 @@ def sympy_interp(
     except Exception:
         log.warning("failed while executing %s(%s)", handler_name, args)
         raise
+
```
refactor so it's easier to hash cons without having this file depend on dict/nodes

pianpwk committed:
comments

pianpwk committed:
serdes expected failure

pianpwk committed:
fix wrong import

ezyang commented on test/dynamo/test_unspec.py:
```diff
@@ -248,6 +248,8 @@ def fn(x, scaler):
         self.assertEqual(ref.device, res.device)
 
+    @unittest.expectedFailure
+    # can't turn ToFloat(Proxy(s1)) intermediate size call into compute
```
Err, this just seems like a regression right?

pianpwk replied:
@ezyang Hmm, what would be the right way to reify the last 2 size calls here?
```
def forward(self, s0 : torch.SymInt, s1 : torch.SymInt, s2 : torch.SymInt, L_image_ : torch.Tensor):
    l_image_ = L_image_
    getitem = l_image_[None]
    interpolate = torch.nn.functional.interpolate(getitem, size = None, scale_factor = 1.873536229133606, mode = 'bilinear', recompute_scale_factor = True, align_corners = False)
    image = interpolate[0]
    size = image.size()
    getitem_2 = size[0]  # shape: s0
    getitem_3 = size[1]  # shape: TruncToInt(1.87353622913361*ToFloat(s1))
    getitem_4 = size[2]  # shape: TruncToInt(1.87353622913361*ToFloat(s1))
    return (getitem_2, getitem_3, getitem_4)
```

I think the issue is maybe with _sympy/interp(), the ToFloat(s1) call gets handled by PythonReferenceAnalysis as a basic Python `float(s1)` call which leads to the error.

ezyang replied:
Ah, make it call torch.sym_float rather than float

pianpwk replied:
Ah I see, thank you! That worked, but now I'm seeing a downstream error with TruncToInt, where the input is a Proxy, and the [check](https://github.com/pytorch/pytorch/blob/e9aefad641f24072f36badf4b221ebca3f222966/torch/utils/_sympy/functions.py#L655) `if number in (sympy.oo, int_oo):` tries to go to control flow tracing and fails:
```
  File "/data/users/pianpwk/pytorch/torch/utils/_sympy/functions.py", line 655, in eval
    if number in (sympy.oo, int_oo):
  File "/data/users/pianpwk/pytorch/torch/fx/proxy.py", line 447, in __bool__
    return self.tracer.to_bool(self)
  File "/data/users/pianpwk/pytorch/torch/fx/proxy.py", line 307, in to_bool
    raise TraceError('symbolically traced variables cannot be used as inputs to control flow')
torch._dynamo.exc.InternalTorchDynamoError: symbolically traced variables cannot be used as inputs to control flow
```

The full expression here is `TruncToInt(1.87353622913361*ToFloat(s1))`

Would you know how to change the handler or sympy functions to work well with Proxys?

ezyang replied:
You shouldn't get here with proxies. Everything in PythonReferenceAnalysis must be traceable. If you look further up your call stack you should find out which method needs to be overridden.

pianpwk replied:
Oh, I think it's because PythonReferenceAnalysis doesn't implement this, it goes all the way to ReferenceAnalysis.

pianpwk replied:
I guess one issue is, when we're converting a shape call on an intermediate tensor into compute on input shapes, we're checking whether all expression symbols are tracked & we're operating any on non-placeholder nodes, but we're unaware if PythonReferenceAnalysis supports reifying the call - this happens with other calls like int_truediv too.

ezyang replied:
Everything is supposed to be supported. Anything that's missing is a bug :o)

pianpwk committed:
Merge branch 'main' of https://github.com/pytorch/pytorch into pianpwk/dedup_runtime_asserts

pianpwk commented on torch/utils/_sympy/interp.py:
```diff
@@ -106,39 +106,14 @@ def handlers():
-    elif isinstance(expr, sympy.Symbol):
-        return env[expr]
-
+def run_sympy_handler(analysis, args, expr, index_dtype=torch.int64):
```
Ah this is unintended, I'll make it private.

pianpwk commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -48,31 +57,30 @@ def insert_deferred_runtime_asserts(
+
+        # turns into ->
+        s0 = x.shape[0]
+        s1 = y.shape[0]
```
I see, I guess I've been too used to export graphs rather than dynamo graphs, they typically get projected out there.

pianpwk committed:
address comments, op count tests

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -85,60 +93,140 @@ def insert_deferred_runtime_asserts(
     symbol_to_proxy: Dict[sympy.Symbol, fx.Proxy] = {}
     placeholders = set()
-    last_placeholder = None
+    first_non_placeholder = next(iter(graph.nodes))
```
This value is never used, since every graph must have a return node to be well formed, so we are guaranteed to find a non-placeholder node. So you could initialize this to None, and then assert is not None if you need to shutup the typechecker later.

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -85,60 +93,140 @@ def insert_deferred_runtime_asserts(
+                torch.ops.aten.sym_size.int,
+                torch.ops.aten.sym_stride.int,
+                torch.ops.aten.storage_offset.default,
+            )
```
I actually think you can probably do some more generic test here. Specifically, if I have a function which returns a SymInt, and whose inputs contain a Tensor, THEN I should try to recompute from input shapes. This covers all of the functions here, but really there is nothing special about these functions, any function that returns a SymInt ought to be eligible.

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -85,60 +93,140 @@ def insert_deferred_runtime_asserts(
+        ):  # export: sym_size.int(node, dim)
+            val = _get_sym_val(node)
+        if (
+            node.target == operator.getitem
```
The getitem test here would also be subsumed, since getitem on a tuple to extract a single SymInt also would go the `get_sym_val` path.

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -85,60 +93,140 @@ def insert_deferred_runtime_asserts(
+        )
 
     # Identify what symbols we need to reify.  This isn't strictly needed
     # but helps reduce churn on the graph
```
It might be simpler to just not identify what symbols we need to reify, and unconditionally reify all of them

pianpwk replied:
I see, I think dynamo more nicely handles dead code than export after this pass happens, export will leave these unused nodes in the graph. But I'll add this at the end to delete unused reified symbols, this should be simpler than tracking needed_symbols now:
```
    # delete unused reified symbols
    for expr, proxy in symbol_to_proxy.items():
        if (
            isinstance(expr, sympy.Symbol)
            and not proxy.node.users
        ):
            log.debug("deleting unused reified symbol for %s", expr)
            gm.graph.erase_node(proxy.node)
```

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -85,60 +93,140 @@ def insert_deferred_runtime_asserts(
 
+    # Track asserts/checks we've added
+    added_asserts: Set["sympy.Expr"] = set()
+    constrained_unbacked_symbols: Set["sympy.Symbol"] = set()
```
Since we have a hash cons going on, it's not entirely clear why we need these extra sets. We'll see, I guess.

pianpwk replied:
I guess it's because they're tracking _check/_assert_scalar or constrain_range calls on top of nodes with symbolic expressions, and the assert/constrain nodes themselves don't have attached expressions.

I guess since the hash cons already stores "signatures", these can be tracked as ("assert", expr) or ("sym_constrain_range", expr)

ezyang replied:
Yeah, I think I was convinced that this is basically fine.

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -85,60 +93,140 @@ def insert_deferred_runtime_asserts(
+        from torch.utils._sympy.interp import _run_sympy_handler, sympy_interp
+
+        # base cases, don't cache
+        if isinstance(expr, (Integer, Number, Symbol, BooleanAtom)):
```
To be clear, the symbol "caches" because the handler just consults symbol_to_proxy. Not sure what happens with the numbers but I guess they get inlined as literals into FX so they don't need to be hash consed.

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -85,60 +93,140 @@ def insert_deferred_runtime_asserts(
+            )
+        # hash cons
+        if expr in symbol_to_proxy:
+            return symbol_to_proxy[expr]
```
symbol_to_proxy kind of a misnomer here right? Since you're storing exprs in the key too

pianpwk replied:
renaming to expr_to_proxy

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -85,60 +93,140 @@ def insert_deferred_runtime_asserts(
+        # hash cons on arguments
+        fx_args = [
+            (
+                symbol_to_proxy[arg]
```
You don't actually need to do this, because the recursive call to `_sympy_interp` will hit the hash cons

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -85,60 +93,140 @@ def insert_deferred_runtime_asserts(
+    def _is_bound_expr_for_symbol(expr: "sympy.Expr") -> bool:
+        # This is probably unnecessary, but since torch._check() calls for single-symbol bounds
+        # like u0 >= 0, 10 >= u0 accumulate range info in the ShapeEnv, and we insert sym_constrain_range calls
+        # anyways, we designate these calls as redundant and remove them.
```
I... sort of see what you're going for here? But I'm not exactly sure how you're doing it / maybe there is a better way.

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -85,60 +93,140 @@ def insert_deferred_runtime_asserts(
+                    and _is_bound_expr_for_symbol(ra.expr)
+                )
+                # if we've already added a constrain_range call for this symbol,
+                # then single-symbol bound asserts like u0 >= 0, u0 <= 5 are redundant.
```
I think my preference is that constrain ranges always turn into a single symbol assert like u0 >= 0, then everything is covered by `added_asserts`

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -222,89 +306,171 @@ def match_symbol(symint, cb):
+                ):
+                    gm.graph.erase_node(node)
+                else:
+                    added_asserts.add(assert_expr)
```
good

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -222,89 +306,171 @@ def match_symbol(symint, cb):
+                else:
+                    added_asserts.add(assert_expr)
+
+            # hash cons on symbolic values
```
would be clearer to say something like "replace function calls which return SymInt with direct references to FX nodes built up from the sympy expression"

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -222,89 +306,171 @@ def match_symbol(symint, cb):
+                    or (
+                        _is_intermediate_tensor_shape_call(node)
+                        and not (sym_expr.free_symbols - symbol_to_proxy.keys())
+                    )  # shape call on intermediate tensor, turn into computation on input shapes
```
Yeah, so concretely, I would just suggest unconditionally generating the sympy expression (hitting memo if necessary) when you get here, instead of gating on the logic here

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -222,89 +306,171 @@ def match_symbol(symint, cb):
+                    sym_expr in symbol_to_proxy  # example value is redundant
+                    or (
+                        _is_intermediate_tensor_shape_call(node)
+                        and not (sym_expr.free_symbols - symbol_to_proxy.keys())
```
This is a bit suspicious to me, this condition should always be true, unless this is an unbacked binding, but we were supposed to treat unbacked bindings specially anyway

pianpwk replied:
Yeah, I guess this handles things downstream of unbacked bindings, and it's partly an issue where AOTAutograd produces unbacked binding symbol keys that don't match the example values (e.g. u1 vs. u6), so we can't reify properly.

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -222,89 +306,171 @@ def match_symbol(symint, cb):
+                        symbol_to_proxy[sym_expr] = _sympy_interp(symbol_to_proxy, sym_expr)
+                        # won't try DCE-ing tensor compute here
+                    hash_node = symbol_to_proxy[sym_expr].node
+                    node.replace_all_uses_with(hash_node)
```
this is smart

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -222,89 +306,171 @@ def match_symbol(symint, cb):
+            if not any(ras for ras in ras_by_symbol.values()):
+                # if no runtime asserts are present, only do CSE,
+                # don't proceed further and add refinement calls
+                continue
```
Not sure I understand this?

pianpwk replied:
I guess this was more to leave things in the current state, where the pass is skipped if runtime_asserts is empty, and no constrain size calls are added. Otherwise I run into dynamo export issues with test cases like this:
```
def test_sym_contains(self):
    def f(x, y):
        return x.size(0) in y

    gm, _ = torch._dynamo.export(f, aten_graph=True)(torch.ones(2), torch.ones(3))
```

And we end up with this graph, which errors with a "constraining non-symbols NYI" on the sym_constrain_range call:
```
def forward(self, L_x_ : torch.Tensor, L_y_ : torch.Tensor):
    l_x_ = L_x_
    l_y_ = L_y_
    sym_size_int = torch.ops.aten.sym_size.int(l_x_, 0);  l_x_ = None
    eq = torch.eq(l_y_, sym_size_int);  l_y_ = sym_size_int = None
    any_1 = torch.any(eq);  eq = None
    item = any_1.item();  any_1 = None
    cast_symbool_to_symint_guardless = torch.fx.experimental.symbolic_shapes.cast_symbool_to_symint_guardless(item)
    sym_constrain_range_default = torch.ops.aten.sym_constrain_range.default(cast_symbool_to_symint_guardless, min = 0, max = 1);  cast_symbool_to_symint_guardless = None
    return (item,)
```

If we switch the constrain_range call to u0 >= 0 and u0 <= 1 assertions, I see this error instead:
```
RuntimeError: Piecewise((1, Eq(u2, 1)), (0, True)) is not tracked with proxy for <torch.fx.experimental.proxy_tensor.PythonKeyTracer object at 0x7fdfbc2c9c00>

While executing %_assert_scalar_default : [num_users=0] = call_function[target=torch.ops.aten._assert_scalar.default](args = (%ge, Runtime assertion failed for expression u0 >= 0 on node 'ge'), kwargs = {})
Original traceback:
None
```

I'm not sure how to handle converted sym bools like this?

ezyang replied:
OK, I don't fundamentally object to this, but the errors you reported here are legitimate problems that we eventually do need to fix.

As for the problem here, at a guess `cast_symbool_to_symint_guardless` isn't tracing correctly with proxies.

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -222,89 +306,171 @@ def match_symbol(symint, cb):
+                continue
+
+            # If the symbol is used, we'll call sym_constrain_range(_for_size) later when we see it anyways,
+            # so delete calls before that
```
Hmm, but aren't we obligated to constrain symbols as size, even if they are otherwise not used anywhere in the FX graph?

pianpwk replied:
Hm I wasn't sure, in main if there aren't any torch._check_is_size calls, then no sym_constrain_range calls show up in the graph. For example this test case:
```
class M1(torch.nn.Module):
    def forward(self, x1, x2, x3, y):
        z1 = x1.item()
        z2 = x2.item()
        z3 = x3.item()
        torch._check(z1 == (z2 + z3))
        if z2 + z3 == z1:
            return y * 2
        else:
            return y + 3
```

ezyang replied:
I suppose you are threading the needle here. A typically situation something is marked size-like without a `_check_is_size` call in the graph, is if the size is passed into a function which internally, in its meta, checks that it is a size. The near miss here is that, I suppose you would argue, then the size must have been used by a function, so it cannot be dead. I suppose I could still synthetically cause this situation if later I DCE'd the function in question. Another way something can be marked size-like is if you have something that runs in Dynamo that marks the symbol as size-like, but that marking doesn't make it into the graph.

At the end of the day, this is about invariants. If you assume that all size-like variables must be used in the body of the graph, you have forced the rest of the system to actually uphold this invariant. But there isn't really any good reason to require this, and not much more complicated for the pass to do the right thing in the absence of the invariant. The trade seems clear to me.

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -222,89 +306,171 @@ def match_symbol(symint, cb):
                 for s, keypath in unbacked_bindings.items():
                     defs.append(s)
-
-                    # TODO: some CSE when generating these nodes can probably
```
Actually, it's not clear to me that you actually implemented the CSE that was specified here. We're talking about CSE when there is something like tensors[0].size(0) is u0, tensors[0].size(1), it could be profitable to CSE tensors[0]. I don't think this is a big deal but I think the comment should stay.

pianpwk replied:
Hmm, would you know of a good test case to try this out? I haven't really found export/dynamo test cases where the keypath length is > 1, so struggling to debug this

ezyang replied:
I'd probably just leave the comment in here / clarify

ezyang reviewed (approved):
lgtm

pianpwk committed:
runtime_assert changes, exported program signature, tests

pianpwk committed:
_trace.py

pianpwk committed:
merge conflict

pianpwk committed:
comments

pianpwk committed:
Merge branch 'main' of https://github.com/pytorch/pytorch into pianpwk/dedup_runtime_asserts

pianpwk committed:
failing tests, lint

pianpwk committed:
syntax

pianpwk committed:
reify

pianpwk committed:
reify

pianpwk committed:
reify

pianpwk committed:
stop trying to be smart

pianpwk committed:
runtime asserts

pianpwk committed:
testfail

pianpwk committed:
don't reify sympy.Functions, turn off asserts for onnx export

pianpwk committed:
remove expected fail

pianpwk committed:
some more tests, silence onnx runtime asserts

pianpwk committed:
recursively check for sympy.Function, lint, tests

pianpwk committed:
Merge branch 'main' of https://github.com/pytorch/pytorch into pianpwk/dedup_runtime_asserts

pianpwk committed:
hopefully last lint, last test failure

pianpwk committed:
random action.yaml failures, unintended comment out

pianpwk committed:
[runtime asserts] deduplicate runtime asserts & CSE (#128599)

Summary:
This PR adds deduplication and CSE for runtime asserts. Existing size computation in the graph is CSE'd along with added runtime asserts, and redundant asserts are removed. Shape calls on intermediate tensors are also turned into compute on input sizes if possible, allowing intermediate tensors to be freed earlier. For example:
```
z = torch.cat([x, x], dim=0)  # 2*s0
w = z.repeat(y.shape[0])  # 2*s0*s1
_w = w.shape[0]
# something with _w ...

# turns into ->
s0 = x.shape[0]
s1 = y.shape[0]
_w0 = 2 * s0
_w = _w0 * s1
```

Additionally, constrain_range calls are deduplicated. Single-symbol bound checks for unbacked symbols (e.g. u0 >= 0, u0 <= 5) and sym_constrain_range.default calls are also removed, since they accumulate range info in the ShapeEnv, and are replaced with two _assert_scalar.default calls that check the min/max bounds. For example:
```
torch.sym_constrain_range_for_size(n, min=2, max=16)
torch.sym_constrain_range(n, min=4, max=20)
torch._check(n >= 0)
torch._check(n >= 3)
torch._check(n <= 14)

# turns into
torch.sym_constrain_range_for_size(n)
torch._check(n >= 4)
torch._check(n <= 14)
```



cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang


Reviewed By: angelayi

Differential Revision: D58513724

Pulled By: pianpwk

ezyang commented:
Just checking in to see if things are alright or if there are problems

pianpwk commented:
@ezyang yep, I was cleaning up some internal test failures, just attempted to land. A merge conflict popped up but it should be in soon.

pianpwk committed:
merge conflict #1

pianpwk committed:
merge conflict #2

pianpwk commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but it seems to cause a landrace and fails inductor/test_cudagraph_trees in trunk https://hud.pytorch.org/pytorch/pytorch/commit/0267b2ddcb58aa66b2b62336216da7df4f9939d8' -c landrace

I'm reverting the change to see if it restores trunk sanity. I'll reland this if it proves to be unrelated

pianpwk committed:
Merge branch 'main' of https://github.com/pytorch/pytorch into pianpwk/dedup_runtime_asserts

pianpwk committed:
operator.le/ge

pianpwk commented:
@huydhn I fixed the cudagraph failure and it's not showing up locally or on CI, will try to merge again

pianpwk commented:
@pytorchbot merge

izaitsevfb commented:
@pytorchbot revert -m "breaking internal APS tests, see D59498864" -c ghfirst

pianpwk commented:
merged in https://github.com/pytorch/pytorch/pull/130380 instead after some codev issues, closing this.
